pvlib/iotools/sodapro.py

```diff
@@ -0,0 +1,368 @@
+    --------
+    pvlib.iotools.read_cams, pvlib.iotools.get_cams
+
+    References
+    ----------
+    .. [1] `CAMS Radiation Service Info
+       <http://www.soda-pro.com/web-services/radiation/cams-radiation-service/info>`_
+    .. [2] `CAMS McClear Service Info
+       <http://www.soda-pro.com/web-services/radiation/cams-mcclear/info>`_
+    """
+    meta = {}
+    # Initial lines starting with # contain meta-data
+    while True:
+        line = fbuf.readline().rstrip('\n')
+        if line.startswith('# Observation period'):
+            # The last line of the meta-data section contains the column names
+            names = line.lstrip('# ').split(';')
+            break  # End of meta-data section has been reached
+        elif ': ' in line:
+            meta[line.split(': ')[0].lstrip('# ')] = line.split(': ')[1]
+
+    # Convert latitude, longitude, and altitude values from strings to floats
+    for k_old in list(meta.keys()):
+        k_new = k_old.lstrip().split(' ')[0].lower()
+        if k_new in ['latitude', 'longitude', 'altitude']:
+            meta[k_new] = float(meta.pop(k_old))
+
+    meta['radiation_unit'] = {True: 'Wh/m^2', False: 'W/m^2'}[integrated]
+
+    # Determine the time_step from the meta-data dictionary
+    time_step = SUMMATION_PERIOD_TO_TIME_STEP[
+        meta['Summarization (integration) period']]
+    meta['time_step'] = time_step
+
+    data = pd.read_csv(fbuf, sep=';', comment='#', header=None, names=names)
+
+    obs_period = data['Observation period'].str.split('/')
+
+    # Set index as the start observation time (left) and localize to UTC
+    if (label == 'left') | ((label is None) & (time_step != '1M')):
+        data.index = pd.to_datetime(obs_period.str[0], utc=True)
+    # Set index as the stop observation time (right) and localize to UTC
+    # default label for monthly data is 'right' following Pandas' convention
+    elif (label == 'right') | ((label is None) & (time_step == '1M')):
+        data.index = pd.to_datetime(obs_period.str[1], utc=True)
+
+    # Change index for time_step '1d' and '1M' to be date and not datetime
+    if (time_step == '1d') | (time_step == '1M'):
+        data.index = pd.DatetimeIndex(data.index.date)
+    # For monthly data with 'right' label, the index should be the last
+    # date of the month and not the first date of the following month
+    if (time_step == '1M') & (label != 'left'):
+        data.index = data.index - pd.Timedelta(days=1)
+
+    if not integrated:  # Convert radiation values from Wh/m2 to W/m2
+        integrated_cols = [c for c in CAMS_INTEGRATED_COLUMNS
+                           if c in data.columns]
+
+        if time_step == '1M':
+            time_delta = (pd.to_datetime(obs_period.str[1])
+                          - pd.to_datetime(obs_period.str[0]))
+            hours = time_delta.dt.total_seconds()/60/60
+            data[integrated_cols] = data[integrated_cols].\
+                divide(hours.tolist(), axis='rows').round(4)
```

Comment:
Is `tolist()` necessary here?  And why `.round(4)`?

Reply:
I have removed the .round(4) and added check_less_precise=True to assert_frame_equal. This works nicely, without requiring any additional modifications. :D

The .tolist() is necessary, I believe it's because the dataframe and series does not have the same index. Are you interested in an alternative solution?

Reply:
Ah I see, I agree that there is an index mismatch.  I don't immediately see an alternative that is any cleaner, so `.tolist()` is fine with me 👍